torch.empty_like and torch.zeros_like raise error if any memory format is provided with sparse input (#43699)

vfdev-5 opened the pull request:
Fixes #43699 

- Changed the order of `TORCH_CHECK` and `if (options.layout() == kSparse && self.is_sparse())`
inside `empty_like` method.

- [x] Added tests

EDIT: 

More details on that and why we can not take zeros_like  approach.
Python code :
```python
res = torch.zeros_like(input_coalesced, memory_format=torch.preserve_format)
```
is routed to
```c++
// TensorFactories.cpp
Tensor zeros_like(
    const Tensor& self,
    const TensorOptions& options,
    c10::optional<c10::MemoryFormat> optional_memory_format) {
  if (options.layout() == kSparse && self.is_sparse()) {
    auto res = at::empty({0}, options); // to be resized
    res.sparse_resize_and_clear_(
        self.sizes(), self.sparse_dim(), self.dense_dim());
    return res;
  }
  auto result = at::empty_like(self, options, optional_memory_format);
  return result.zero_();
}
```
and passed to `if (options.layout() == kSparse && self.is_sparse())`

When we call in Python
```python
res = torch.empty_like(input_coalesced, memory_format=torch.preserve_format)
```
it is routed to
```c++
Tensor empty_like(
    const Tensor& self,
    const TensorOptions& options_,
    c10::optional<c10::MemoryFormat> optional_memory_format) {
  TORCH_CHECK(
    !(options_.has_memory_format() && optional_memory_format.has_value()),
    "Cannot set memory_format both in TensorOptions and explicit argument; please delete "
    "the redundant setter.");
  TensorOptions options =
      self.options()
          .merge_in(options_)
          .merge_in(TensorOptions().memory_format(optional_memory_format));
  TORCH_CHECK(
      !(options.layout() != kStrided &&
          optional_memory_format.has_value()),
      "memory format option is only supported by strided tensors"); 
  if (options.layout() == kSparse && self.is_sparse()) {
    auto result = at::empty({0}, options); // to be resized
    result.sparse_resize_and_clear_(
        self.sizes(), self.sparse_dim(), self.dense_dim());
    return result;
  }
```

cc @pearu

Commit:
Fixes #43699
- Changed the order of TORCH_CHECK and if (options.layout() == kSparse && self.is_sparse())
inside empty_like method

Commit:
Fixed flake and failing test

pearu commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -282,17 +282,18 @@ Tensor empty_like(
           .merge_in(options_)
           .merge_in(TensorOptions().memory_format(optional_memory_format));
```
When `options.layout()` is `kSparse`, I think we should avoid adding `memory_format` to `options` because it is invalid for non-strided tensors.

vfdev-5 replied:
Thanks for the comment, OK I see. Let's bypass using `optional_memory_format`  as in C++ `zeros_like`.

pearu reviewed (changes requested):
LGTM but with a nit that one should avoid setting `memory_format` option when the input tensor is not a strided tensor.

Commit:
Updated code and tests according to the review

pearu commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -277,18 +277,24 @@ Tensor empty_like(
+    if (optional_memory_format.has_value()) {
+      TORCH_CHECK(
+        optional_memory_format == MemoryFormat::Preserve,
+        "This memory format option is not supported by sparse tensors");
```
I wonder if we could improve the exception message to be more informative.
For instance, use
```
TORCH_CHECK(optional_memory_format == MemoryFormat::Preserve,
            "Memory format for sparse tensors must be Preserve, got ", optional_memory_format);

Commit:
Fixed failing test

vfdev-5 committed:
Update TensorFactories.cpp

pearu commented on test/test_sparse.py:
```diff
@@ -1619,6 +1619,75 @@ def test_shape(i_shapes, v_shapes, nnzs):
+            )
+            _test_empty_like(input_coalesced)
+
+        if self.is_uncoalesced:
```
nit: `else:`

pearu reviewed (approved):
LGTM, have a minor nit.
Try to rebase to fix some CI failures regarding merge conflicts.

Commit:
Merge branch 'master' of https://github.com/pytorch/pytorch into vfdev-5/sparse-empty-like-mem-preserve

mruberry reviewed (approved):
Looks great and thorough!

One question I have about sparse testing as we move forward is if we expect test_sparse.py to always be a thing? PyTorch has a lot of functions, so test_sparse.py being around in its current form suggests one or both of the following will be true:

- not many functions will support sparse values
- sparse functions will always be different from other torch functions, possibly so different that we should have `torch.sparse.foo` instead of `torch.foo` for functions operating on sparse tensors

I imagine they'll always be a need for test_sparse.py just like there's a need for test_cuda.py, but as time has gone on the tests in test_cuda.py have focused more on CUDA-specific components, while CUDA variants of functions are just tested automatically now using the device generic test framework. 

How we write the tests may not seem very important (except as it impacts development time), but I think the tests can be helpful in thinking about and enforcing the consistency of our designs. Certainly not an issue for this PR, however!

mruberry reviewed (changes requested):
Update: after discussion with @ngimel, we'd like to change this so that when a memory_format is specified we error out as before. That is, we would like to not support setting any memory format for sparse tensors and continue to throw the "memory format option is only supported by strided tensors" in those cases. That is, both zeros_like and empty_like should error out if this argument is provided. The current behavior of empty_like is correct.

vfdev-5 commented:
@mruberry thanks for the review and the update ! 

> That is, we would like to not support setting any memory format for sparse tensors and continue to throw the "memory format option is only supported by strided tensors" in those cases. That is, both zeros_like and empty_like should error out if this argument is provided. The current behavior of empty_like is correct.

I agree that the formats like `torch.channels_last, torch.channels_last_3d, torch.contiguous_format` have not much meaning for sparse tensors. I thought that we can tolerate `torch.preserve_format` for sparse tensors.
I'd like to detail the context of this PR and discuss what can be done in this case. There are several functions that call `zeros_like` op with memory preserve option to construct its derivative. For example, `sign` op : 
https://github.com/pytorch/pytorch/blob/2b70f827373fe5c90e15ce7ba0a1d331dcd57379/tools/autograd/derivatives.yaml#L912-L913

Calling `zeros_like` internally redirected to `empty_like` with `at::MemoryFormat::Preserve` as option and thus raises a runtime error.
If we would like to have both zeros_like and empty_like raising an error if memory format argument is provided, than we need to deal with derivatives... What do you think ?

mruberry commented:
> @mruberry thanks for the review and the update !
> 
> > That is, we would like to not support setting any memory format for sparse tensors and continue to throw the "memory format option is only supported by strided tensors" in those cases. That is, both zeros_like and empty_like should error out if this argument is provided. The current behavior of empty_like is correct.
> 
> I agree that the formats like `torch.channels_last, torch.channels_last_3d, torch.contiguous_format` have not much meaning for sparse tensors. I thought that we can tolerate `torch.preserve_format` for sparse tensors.
> I'd like to detail the context of this PR and discuss what can be done in this case. There are several functions that call `zeros_like` op with memory preserve option to construct its derivative. For example, `sign` op :
> https://github.com/pytorch/pytorch/blob/2b70f827373fe5c90e15ce7ba0a1d331dcd57379/tools/autograd/derivatives.yaml#L912-L913
> 
> Calling `zeros_like` internally redirected to `empty_like` with `at::MemoryFormat::Preserve` as option and thus raises a runtime error.
> If we would like to have both zeros_like and empty_like raising an error if memory format argument is provided, than we need to deal with derivatives... What do you think ?

Thanks for the clear summary. You make a compelling case. Let me discuss with @ngimel and we'll get back to you shortly. 

Are there other places in the code or in our documentation where we discuss the relationship between sparse tensors and memory formats?

ngimel commented:
For `sign` derivative (and most likely for others) the correct solution is to remove memory format specification from derivative formula. No value of memory format makes sense for sparse tensors, so allowing `Preserve` only is very confusing. What does it preserve?

vfdev-5 commented:
@mruberry @ngimel thanks for your comments ! I agree that memory format makes sense for strided tensors only. Sparse layouts (coo, csr) use strided tensors, and strided tensors may use different memory formats. 

For this PR, let me do what you suggest: 
- zeros_like and empty_like should error for sparse tensor if any optional memory format is provided
- remove memory format specification from derivative formula

Commit:
Merge branch 'master' of https://github.com/pytorch/pytorch into vfdev-5/sparse-empty-like-mem-preserve

Commit:
- Updated cpp code
- Updated tests

mruberry commented on test/test_torch.py:
```diff
@@ -14581,10 +14578,6 @@ def test_helper(x, memory_format):
             self.assertFalse(like.is_contiguous())
             self.assertTrue(like.is_contiguous(memory_format=memory_format))
 
-            sparse = x.to_sparse()
```
This (sub)test can be added back, right?

vfdev-5 replied:
yes, you're right. I somehow missed that. I'll get it back. Thanks !

mruberry commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -860,6 +860,9 @@ Tensor zeros_like(
     const TensorOptions& options,
     c10::optional<c10::MemoryFormat> optional_memory_format) {
   if (options.layout() == kSparse && self.is_sparse()) {
```
This is interesting as the check is that both the requested layout and the example tensor be sparse.

I tried creating a sparse tensor on Colab from a dense example:

```
t = torch.tensor((1, 2, 3))
torch.zeros_like(t, layout=torch.sparse_coo)
: Could not run 'aten::empty_strided' with arguments from the 'SparseCPU' backend. 'aten::empty_strided' is only available for these backends: [CPU, CUDA, BackendSelect, Autograd, Profiler, Tracer].
```

and a dense tensor from a sparse example:

```
t = torch.tensor((1, 2, 3)).to_sparse()
torch.zeros_like(t, layout=torch.strided, memory_format=torch.preserve_format)
: tensor([0, 0, 0])
```

and a sparse tensor from a sparse example:

```
t = torch.tensor((1, 2, 3)).to_sparse()
result = torch.zeros_like(t)
: tensor(indices=tensor([], size=(1, 0)),
       values=tensor([], size=(0,)),
       size=(3,), nnz=0, layout=torch.sparse_coo)
```

If the user doesn't specify a layout explicitly, what is layout set to? In particular, should this check `(has_layout() && layout == kSparse) || (!has_layout() && self.is_sparse())`?

vfdev-5 replied:
> If the user doesn't specify a layout explicitly, what is layout set to? 

I think, the `options` is derived from `self` if no layout is specified in python
```
# self: sparse 
# option layout: default
t = torch.tensor((1, 2, 3)).to_sparse()
torch.zeros_like(t)
> tensor(indices=tensor([], size=(1, 0)),
       values=tensor([], size=(0,)),
       size=(3,), nnz=0, layout=torch.sparse_coo)

# self: dense 
# option layout: default
t = torch.tensor((1, 2, 3))
torch.zeros_like(t)
> tensor([0, 0, 0])
```
this is due to how the code is generated : 
```
# torch/csrc/autograd/generated/python_torch_functions.cpp
#    │21490     // aten::zeros_like(Tensor self, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
#    │21491     auto self = _r.tensor(0);
#    │21492     const auto options = TensorOptions()
#    │21493         .dtype(_r.scalartypeWithDefault(2, self.scalar_type()))                                                                                                                #    │21494         .device(_r.deviceWithDefault(4, self.device()))                                                                                                                        #    │21495         .layout(_r.layoutWithDefault(3, layout_from_backend(self.options().backend())))                                                         #    │21496         .requires_grad(_r.toBool(6))                                                                                                                                           #    │21497         .pinned_memory(_r.toBool(5));                                                                                                                                          #    │21498     torch::utils::maybe_initialize_cuda(options);                                                                                                                              #    │21499     auto dispatch_zeros_like = [](const Tensor & self, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) -> Tensor {
#    │21500       pybind11::gil_scoped_release no_gil;                                                                                                                                     #    │21501       return torch::zeros_like(self, options, memory_format);                                                                                                                  #    │21502     };                                                                                                                                                                         #   >│21503     return wrap(dispatch_zeros_like(self, options, _r.memoryformatOptional(1)));                                                                                               #    │21504     Py_RETURN_NONE;                                                                                                                                                            #    │21505     END_HANDLE_TH_ERRORS                                                                                                                                                       #    │21506   }
```
So, maybe we can keep the current check.

mruberry commented on test/test_sparse.py:
```diff
@@ -1611,6 +1611,74 @@ def test_shape(i_shapes, v_shapes, nnzs):
+        sparse_tensor, _, _ = self._gen_sparse(len([2, 3, 4]), 9, [2, 3, 4] + [3, 4, 5, 6])
+        for mem_format in [torch.channels_last, torch.channels_last_3d, torch.contiguous_format, torch.preserve_format]:
+            with self.assertRaisesRegex(RuntimeError, "memory format option is only supported by strided tensors"):
+                result = torch.zeros_like(sparse_tensor, memory_format=mem_format)
```
I would add tests for zeros_like with a dense tensor but sparse layout request, and for a sparse tensor but a strided layout request. The first case should also error (although, as per a previous comment not for any memory-format-related reason) and the second case should work with memory_format. 

Ensuring zeros_like(dense -> sparse) errors will also ensure this test is updated properly if we ever change this code.

mruberry commented on test/test_sparse.py:
```diff
@@ -1611,6 +1611,74 @@ def test_shape(i_shapes, v_shapes, nnzs):
+            self.assertEqual(result.sparse_dim(), sparse_tensor.sparse_dim())
+            self.assertEqual(result.dense_dim(), sparse_tensor.dense_dim())
+
+            for mem_format in [torch.channels_last, torch.channels_last_3d, torch.contiguous_format, torch.preserve_format]:
```
Same comment as above about extending the test combinations here.

Commit:
Added new tests according to the review

Commit:
Fixed flake8

mruberry commented on test/test_sparse.py:
```diff
@@ -1611,6 +1611,92 @@ def test_shape(i_shapes, v_shapes, nnzs):
+            dense_tensor = sparse_tensor.to_dense()
+            result = torch.zeros_like(dense_tensor, layout=torch.sparse_coo)
+
+        result = torch.zeros_like(sparse_tensor, layout=torch.strided)
```
Just pull these lines into the block where you enumerate the memory formats, here and with empty_like. 

```
sparse_tensor, _, _ = self._gen_sparse(len([2, 3, 4]), 9, [2, 3, 4] + [3, 4, 5, 6])
for mem_format in [torch.channels_last, torch.channels_last_3d, torch.contiguous_format, torch.preserve_format]:
  with self.assertRaisesRegex(RuntimeError, "memory format option is only supported by strided tensors"):
    result = torch.zeros_like(sparse_tensor, memory_format=mem_format)

  result = torch.zeros_like(sparse_tensor, layout=torch.strided, memory_format=mem_format)
  self.assertTrue(result.layout == torch.strided)
```

mruberry reviewed (approved):
Thanks @vfdev-5! I made one minor suggestion about the tests. Just ping me when this is ready.

Commit:
Updated tests
- changed scope of test local functions
- updated tests according to the review

vfdev-5 commented:
@mruberry thanks for the suggestion. I updated the code as you proposed and refactored a bit test_empty_like code.
